test/boost/repair.cc

```diff
@@ -0,0 +1,137 @@
+        , _sst(std::move(sst))
+        , _cfg(std::move(cfg))
+    {
+        _consumer = std::make_unique<upgrading_consumer<std::function<void(mutation_fragment_v2&&)>>>
+            (*_schema, _permit, [this](mutation_fragment_v2&& mf) mutable {
+                _fragments.push_back(std::move(mf));
+            });
+    }
+    
+    virtual future<> push(mutation_fragment mf) override {
+        _consumer->consume(std::move(mf));
+        return make_ready_future();
+    }
+
+    virtual void abort(std::exception_ptr ep) override {}
+
+    virtual void push_end_of_stream() override {
+        flat_mutation_reader_v2 r = make_flat_mutation_reader_from_fragments(_schema, _permit, std::move(_fragments));
+        int estimated_partitions = 1;
+        _sst->write_components(std::move(r), estimated_partitions, _schema, _cfg, encoding_stats{}).get();
+        _sst->load().get();
+    }
+};
+
+mutation_fragment_queue make_test_mutation_fragment_queue(schema_ptr s, reader_permit permit, shared_sstable sst, sstable_writer_config cfg) {
+    return mutation_fragment_queue(std::make_unique<test_mutation_fragment_queue_impl>(std::move(s), std::move(permit), std::move(sst), std::move(cfg)));
+}
+
+class test_repair_writer_impl : public repair_writer::impl {
+    mutation_fragment_queue _queue;
+public:
+    test_repair_writer_impl(mutation_fragment_queue queue)
+        : _queue(std::move(queue))
+    {
+    }
+
+    virtual future<> wait_for_writer_done() override {
+        // Pushing EOS to the underlying mq waits until the sstable is written
+        return make_ready_future();
+    }
+
+    virtual mutation_fragment_queue& queue() override {
+        return _queue;
+    }
+
+    virtual void start_streaming(lw_shared_ptr<repair_writer> writer) override {}
+};
+
+lw_shared_ptr<repair_writer> make_test_repair_writer(schema_ptr schema, reader_permit permit, mutation_fragment_queue mq) {
+    return make_lw_shared<repair_writer>(std::move(schema), std::move(permit), std::make_unique<test_repair_writer_impl>(std::move(mq)));
+}
+
+SEASTAR_TEST_CASE(flush_repair_rows_on_wire_to_sstable) {
+    return test_setup::do_with_tmp_directory([] (test_env& env, sstring tmpdir_path) {
+        tests::reader_concurrency_semaphore_wrapper semaphore;
+        reader_permit permit = env.make_reader_permit();
+        uint64_t seed = tests::random::get_int<uint64_t>();
+        random_mutation_generator gen{random_mutation_generator::generate_counters::no};
+        unsigned long generation{0};
+        schema_ptr s = gen.schema();
+
+        std::vector<mutation> muts = gen(100);
+        auto m = make_lw_shared<replica::memtable>(s);
+        repair_rows_on_wire input;
+
+        for (mutation& mut : muts) {
+            partition_key pk = mut.key();
+            auto m2 = make_lw_shared<replica::memtable>(s);
+            m->apply(mut);
+            m2->apply(mut);
+            flat_mutation_reader reader = downgrade_to_v1(m2->make_flat_reader(s, permit));
```

Comment:
@denesb are we planning to remove this "downgrade_to_v1" stuff in the near future? If we are, let's not write new tests that uses it.

Reply:
This series is part of an effort to move repair off v1 readers. We might be forced to work with v1 readers here for the time being. Of course if possible, we should write all code with v2 readers.

Reply:
We're a bit forced here. The goal is to get rid of queue_reader and queue_reader_handle (both are v1 and we already have a v2 version, but can't use it here yet). Repair pipeline uses them internally on the part that the unit tests cover (repair_rows_on_wire -> repair_row -> flush to disk with some extra steps). The goal of this patchset is to write a unit test for that specific part, so that when I get rid of the v1 version of queue reader (in a subsequent patch) I can replace repair_row with a newly written repair_row_v2 and can skip the downgrades.